ClemBot.Bot/bot/cogs/search_cog.py

```diff
@@ -0,0 +1,159 @@
+            title = related_topics_pattern\
+                .match(topic['FirstURL'])\
+                .group(3)\
+                .replace('_', ' ')
+            title = unquote_plus(title)
+            related_topics_formatted.append(f"[{title}]({topic['FirstURL']})")
+        return '\n'.join(related_topics_formatted)
+
+    def url(self):
+        return self.json_response['AbstractURL']
+
+    def source(self):
+        return self.json_response['AbstractSource']
+
+
+class SearchCog(commands.Cog):
+
+    def __init__(self, bot):
+        self.bot = bot
+
+    @ext.command()
+    @ext.short_help('Searches via DuckDuckGo.')
+    @ext.long_help('Searches via DuckDuckGo\'s instant web API to bring up results from Wikipedia, wikiHow, and more.')
+    @ext.example(['search clemson', 'search computer science', 'search how to tie a tie'])
+    async def search(self, ctx, *, query: str):
+        result = await self.duck_search(query)
+        msg = await self.results(ctx, result) if result.has_result() else await self.no_results(ctx, query)
+        await self.bot.messenger.publish(Events.on_set_deletable, msg=msg, author=ctx.author, timeout=60)
+
+    async def duck_search(self, query: str) -> SearchResult:
+        data = urlencode({
+            "q": query,
+            "format": "json",
+            "pretty": 1,
+            "skip_disambig": 1,
+            "t": "ClemBot"
+        })
+        log.debug(f'Search URL: {search_url}?{data}')
+        async with aiohttp.ClientSession() as session:
+            # redirects should never happen so it's gonna be off by default just in case
+            async with session.get(url=f'{search_url}?{data}', allow_redirects=False) as resp:
+                response = json.loads(await resp.text())
+            log.info(response)
+        return SearchResult(response)
+
+    async def results(self, ctx, result: SearchResult) -> discord.Message:
+        embed = discord.Embed(title=f'[{result.category().name()}] {result.title()} - {result.source()}',
+                              color=Colors.ClemsonOrange,
+                              description=f'{result.abstract()}\n\n**Link**: [{result.title()}]({result.url()})')
+        embed.set_footer(text='Result provided by DuckDuckGo.', icon_url=icon_url)
+        if result.has_related_topics():
+            embed.add_field(name='Related Topics', value=result.related_topics_formatted(), inline=False)
+        if result.has_thumbnail():
+            embed.set_thumbnail(url=result.thumbnail())
+        return await ctx.send(embed=embed)
+
+    async def no_results(self, ctx, query: str) -> discord.Message:
+        embed = discord.Embed(title='No Results', color=Colors.Error,
+                              description=f"Your query '{query}' yielded 0 results.")
+        embed.set_footer(text='Result provided by DuckDuckGo.', icon_url=icon_url)
+        return await ctx.send(embed=embed)
+
+
+def setup(bot):
```

Comment:
move setup to below `category_from_code` for consistency with other cogs reasons